Modal 컴포넌트 구현

bakhacode opened the pull request:
## Issues

- #3

## What is this PR? 👓
  -  라이트한 모달 틀에 각각 다른 스타일과 로직을 가진 두 개의 모달을 주입해서 구현
## Key changes 🔑

## To reviewers 👋
최대한 해본다고 해봤는데 방향이 맞는지는 의문이네요 😂 최대한 카카모토비 코드 참고해가면서 해봤는데 잘못된 부분 있으면 정말 편하게 다 피드백 주셔도 됩니다!

bakhacode committed:
#3 feat: Modal 컴포넌트 구현

- UI 완성
- 핸들러 함수 연결

bakhacode committed:
#3 refactor: 동네 선택, 카테고리 선택 두 개의 모달로 변경

bakhacode committed:
#3 refactor: item 컴포넌트 모달에서 분리

bakhacode committed:
#3 refactor: 모달을 최대한 간단한 컴포넌트로 분리

bakhacode committed:
#3 design: 모달 관련 공통 스타일 분리

bakhacode committed:
#3 style: 경로 대문자 변경

bakhacode committed:
#3 refactor: svg 로딩 방식을 직접 import로 변경

Kakamotobi commented on src/styles/modal/ModalStyles.ts:
이 파일은 `components/common/Modal/` 안으로 옮기는게 더 좋을것 같습니다

bakhacode replied:
고민하다가 스타일로 뺐는데 한 쪽으로 모으는 게 맞군요 수정했습니다!

Kakamotobi commented on src/components/common/Modal/CategoryModal.tsx:
```diff
@@ -0,0 +1,125 @@
+import x from "@assets/icon/x.svg";
```
`x`보다 `xIcon` 또는 `closeIcon`등이 더 명확할거 같습니다

Kakamotobi commented on src/components/common/Modal/CategoryItem.tsx:
```diff
@@ -0,0 +1,24 @@
+    <ModalListItem $isSelected={isSelected} onClick={() => onClick(item.id)}>
+      {item.title}
+      {/* 이미지 확인용 */}
+      {/* <img width="40px" height="40px" src={item.imageUrl} alt={item.title} /> */}
```
확인 다 하셨으면 주석 지워도 될것 같습니다

Kakamotobi commented on src/components/common/Modal/CategoryModal.tsx:
```diff
@@ -0,0 +1,125 @@
+import { useState } from "react";
+
+export default function CategoryModal() {
+  const [selectedCategory, setSelectedCategory] = useState(1);
```
`selectedCategoryId`로 이름을 바꾸면 이해하기 더 쉬울것 같습니다

Kakamotobi commented on src/components/common/Modal/CategoryModal.tsx:
```diff
@@ -0,0 +1,125 @@
+    id: 11,
+    title: "여성의류",
+    imageUrl: "https://i.ibb.co/4fvj6SC/woman-apparel.png",
+  },
+  {
+    id: 12,
+    title: "남성패션/잡화",
+    imageUrl: "https://i.ibb.co/wwfyjyB/man-apparel.png",
+  },
+  { id: 13, title: "게임/취미", imageUrl: "https://i.ibb.co/cwJ74M4/game.png" },
+  {
+    id: 14,
+    title: "뷰티/미용",
+    imageUrl: "https://i.ibb.co/cXrrK0m/beauty.png",
+  },
+  {
+    id: 15,
+    title: "반려동물용품",
+    imageUrl: "https://i.ibb.co/CbwHdNr/pet.png",
+  },
+  { id: 16, title: "도서/음반", imageUrl: "https://i.ibb.co/7WjKkdt/book.png" },
+  {
+    id: 17,
+    title: "티켓,교환권",
+    imageUrl: "https://i.ibb.co/kBhhs2p/ticket.png",
+  },
+  { id: 18, title: "생활", imageUrl: "https://i.ibb.co/T0mnp8m/kitchen.png" },
+  {
+    id: 19,
+    title: "가공식품",
+    imageUrl: "https://i.ibb.co/S0rSyxr/processed-foods.png",
+  },
+  { id: 20, title: "식물", imageUrl: "https://i.ibb.co/rwZhRqh/plant.png" },
+  {
+    id: 21,
+    title: "기타 중고물품",
+    imageUrl: "https://i.ibb.co/tCyMPf5/etc.png",
+  },
+  { id: 22, title: "삽니다", imageUrl: "https://i.ibb.co/g7Gc1w0/buy.png" },
+];
```
Mock data는 msw로 옮기는게 좋을것 같아요 👍

Kakamotobi replied:
제가 MSW 초기화해서 PR 올렸는데 먼저 merge하고 받으신 후에 옮기면 될것 같습니다

bakhacode replied:
넵 Region에 있는 Mock Data도 동일하게 처리하겠습니다!

Kakamotobi commented on src/components/common/Modal/Modal.tsx:
```diff
@@ -0,0 +1,40 @@
+  header: React.ReactNode;
+  body: React.ReactNode;
+};
+export default function Modal({ header, body }: Props) {
+  return (
+    <StyledModal>
+      <ModalHeader>{header}</ModalHeader>
+      <ModalBody>{body}</ModalBody>
+    </StyledModal>
+  );
+}
```
- 모달 구조를 아래처럼하는건 어떨까요?
- 그리고 모달 컴포넌트에서 backdrop(dim)을 적용해주면 좋을거 같아요.
  - i.e. `Modal`은 단순히 `children`을 backdrop앞에 그려주는 컴포넌트.

```tsx
export default function Modal({
  isOpen,
  onClose, // Modal 밖(backdrop)을 클릭스시 닫기 위함.
  children,
}: {
  isOpen: boolean;
  onClose: () => void;
  children: ReactNode;
}) {
  return (
    <StyledModal onClick={onClose}> // Backdrop 적용
      <div>{children}</div> // 모달 박스
    </StyledModal>
  );
}
```
```tsx
// Example 1
<div>
  <Modal isOpen={isRegionModalOpen} onClose={onRegionModalClose}>
    <header>
      <h3>제목</h3>
      <button onClick={onRegionModalClose}>
        <img src={xIcon} alt="동네 설정 모달 닫기" />
      </button>
    </header>
  </Modal>
</div>

// Example 2
<div>
  <Modal isOpen={isRegionModalOpen} onClose={onRegionModalClose}>
    <header>
      <button onClick={onGoBack}>
        <img src={chevronLeft} alt="뒤로 가기" />
      </button>
    </header>
  </Modal>
</div>
```

bakhacode replied:
카카모토비가 제안해주신 방식으로 구조 수정하고 RegionModal 같은 경우 첫 메인 화면과 동네 추가시 화면이 달라져서 분리해봤어요! 나머지 피드백 주신 부분들도 전부 수정완료했습니다 확인하고 수정해야하는 부분 있으면 코멘트 달아주세요! 👍

Kakamotobi commented:
그리고 `components/common/Modal`안에는 `Modal` 범용 컴포넌트 모듈만 들어있는게 좋을거 같아요. `Modal` 컴포넌트를 사용해서 만든 다른 컴포넌트 (Ex: `CategoryModal`)은 따로 빼는게 좋을 것 같습니다 👍

bakhacode committed:
#3 refactor: Modal 구조 변경 및 디렉토리 구조 변경

- Modal을 더 단순한 구조로 변경
- RegionModal을 상태에 따라 달라지는 두 개의 컴포넌트로 분리
- Region과 Category에 속하는 Modal들을 각 폴더로 분리
- 코딩 컨벤션 피드백에 맞게 수정

Kakamotobi commented on src/components/Category/CategoryModal.tsx:
```diff
@@ -1,46 +1,52 @@
-      </IconWrapper>
-    </ModalHeader>
-  );
+  const onCategoryModalClose = () => {
+    //모달 밖(backdrop)을 클릭시 닫기 위한 함수
+  };
```
이거는 props로 받아도 될거 같아요

Kakamotobi commented on src/components/Region/RegionModalAdd.tsx:
```diff
@@ -0,0 +1,98 @@
+  onRegionBack: () => void;
+};
+
+export default function RegionModalAdd({
```
`RegionAddModal`이 더 명확할것 같습니다

Kakamotobi commented on src/components/Region/RegionModalMain.tsx:
```diff
@@ -0,0 +1,117 @@
+  selectedRegionList: string[];
+};
+
+export default function RegionModalMain({
```
`RegionSelectModal`이 좋을것 같습니다

Kakamotobi commented on src/components/Region/RegionModal.tsx:
```diff
@@ -0,0 +1,60 @@
+import RegionModalAdd from "./RegionModalAdd";
+
+export default function RegionModal() {
+  const [isAddMode, setIsAddMode] = useState(false);
```
`isRegionAddModal`이 그나마 좀 더 이해하기 쉬울 것 같아요

Kakamotobi commented on src/components/Region/RegionModalAdd.tsx:
```diff
@@ -0,0 +1,98 @@
+  isAddMode: boolean;
+  isRegionModalOpen: boolean;
+  onRegionModalClose: () => void;
+  onRegionBack: () => void;
```
`onRegionBack` 이름만 봐서는 헷갈리는거 같아요. 🫨

Kakamotobi commented on src/components/Region/RegionModalMain.tsx:
```diff
@@ -0,0 +1,117 @@
+  onRegionAdd,
+  selectedRegionList,
+}: Props) {
+  const canPerformAction = selectedRegionList.length !== 1;
```
읽기 쉽게 아래처럼 해봐도 될것 같습니다.
삭제: 2개일 때만 가능
추가: 1개일 때만 가능

```ts
const selectedOneRegion = selectedRegionList.length === 1
const selectedTwoRegions = selectedRegionList.length === 2
```

bakhacode committed:
#3 refactor: 피드백 반영한 수정사항
